snuba/utils/streams/batching.py

```diff
@@ -61,6 +62,150 @@ class Offsets:
     hi: int  # exclusive
 
 
+@dataclass
+class Batch(Generic[TResult]):
+    results: MutableSequence[TResult] = field(default_factory=list)
+    offsets: MutableMapping[Partition, Offsets] = field(default_factory=dict)
+    created: float = field(default_factory=lambda: time.time())
+    messages_processed_count: int = 0
+    # the total amount of time, in milliseconds, that it took to process
+    # the messages in this batch (does not included time spent waiting for
+    # new messages)
+    processing_time_ms: float = 0.0
+
+
+class BatchProcessor:
+    """
+    The ``BatchProcessor`` is a message processor that accumulates processed
+    message values, periodically flushing them after a given duration of time
+    has passed or number of output values have been accumulated.
+    """
+
+    def __init__(
+        self,
+        consumer: Consumer[TPayload],
+        worker: AbstractBatchWorker[TPayload, TResult],
+        max_batch_size: int,
+        max_batch_time: int,
+        metrics: MetricsBackend,
+    ) -> None:
+        self.__consumer = consumer
+        self.__worker = worker
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+        self.__metrics = metrics
+
+        self.__batch: Optional[Batch[TResult]] = None
+        self.__closed = False
+
+    def process(self, message: Optional[Message[TPayload]]) -> None:
```

Comment:
I don't like this (specifically making this method essentially overloaded on `Message[TPayload]` and `None`) but I can't think of a better way to do it.